cardano-crypto-class/src/Cardano/Crypto/Hash/Blake2b.hs

```diff
@@ -14,24 +17,123 @@ where
 import Control.Monad (unless)
 import Cardano.Crypto.Libsodium.C (c_crypto_generichash_blake2b)
 
-import Cardano.Crypto.Hash.Class (HashAlgorithm, SizeHash, hashAlgorithmName, digest)
+import Cardano.Crypto.Hash.Class (HashAlgorithm (..), SizeHash, hashAlgorithmName, digest)
 import Foreign.Ptr (castPtr, nullPtr)
 import Foreign.C.Error (errnoToIOError, getErrno)
 import GHC.IO.Exception (ioException)
+import qualified Data.ByteString.Short.Internal as SBSI
+import qualified Data.ByteString.Short as SBS
+import Data.ByteString.Short (ShortByteString)
+import Data.Word (Word64)
 
 import qualified Data.ByteString as B
 import qualified Data.ByteString.Internal as BI
+import           Control.DeepSeq (NFData)
+import           NoThunks.Class (NoThunks)
+import GHC.Generics (Generic)
+import qualified Data.Primitive.Types as Prim
+import qualified Data.Primitive.ByteArray as BA
 
 data Blake2b_224
 data Blake2b_256
 
+data Blake2b_224Hash =
+  Blake2b_224Hash
+    {-# UNPACK #-} !Word64
+    {-# UNPACK #-} !Word64
+    {-# UNPACK #-} !Word64
+    {-# UNPACK #-} !Word64
+    {-# UNPACK #-} !Word64
+    {-# UNPACK #-} !Word64
+    {-# UNPACK #-} !Word64
+  deriving (Eq, Ord, Generic)
+
+instance NoThunks Blake2b_224Hash
+instance NFData Blake2b_224Hash
+
+-- | Converts a ShortByteString into a Blake2b_224 hash rep
+--   This will error if the bytestring length is not 28.
+unsafeBlake2b_224HashFromShort :: ShortByteString -> Blake2b_224Hash
+unsafeBlake2b_224HashFromShort sbs =
+   if l == 28
+   then Blake2b_224Hash x0 x1 x2 x3 x4 x5 x6
+   else error $
+         "Attempted to cast bytestring of length "
+      <> show l
+      <> " into Blake2b224 hash rep, but the required length is 28."
+  where
+   l = SBS.length sbs
+   !(SBSI.SBS byteArray) = sbs
+   x0 = Prim.indexByteArray# byteArray 0#
+   x1 = Prim.indexByteArray# byteArray 1#
+   x2 = Prim.indexByteArray# byteArray 2#
+   x3 = Prim.indexByteArray# byteArray 3#
+   x4 = Prim.indexByteArray# byteArray 4#
+   x5 = Prim.indexByteArray# byteArray 5#
+   x6 = Prim.indexByteArray# byteArray 6#
+
+-- | Converts a hash rep into a ShortByteString
+blake2b224HashToShort :: Blake2b_224Hash -> ShortByteString
+blake2b224HashToShort (Blake2b_224Hash x0 x1 x2 x3 x4 x5 x6) = SBSI.SBS byteArray
+  where
+  !(BA.ByteArray byteArray) = BA.byteArrayFromListN 7 [x0,x1,x2,x3,x4,x5,x6]
```

Comment:
Relying on `byteArrayFromListN` here will be a lot more expensive then it has to be because of extra list allocation and an unnecessary loop. Mutable approach is a much better alternative:
```suggestion
  !(BA.ByteArray byteArray) = runST $ do
      mba <- newByteArray 28
      writeByteArray mba 0 x0
      writeByteArray mba 1 x1
      writeByteArray mba 2 x2
      writeByteArray mba 3 x3
      writeByteArray mba 4 x4
      writeByteArray mba 5 x5
      writeByteArray mba 6 x6
      unsafeFreezeByteArray mba
```